TSOClient/tso.common/Utils/TextureUtils.cs

```diff
@@ -659,111 +662,166 @@ private static byte[] GetAlphaIndices(Color[] block, Color minCol, Color maxCol)
-                    }
+                    int dist0 = ColorDistanceSq(c, color0);
+                    int dist1 = ColorDistanceSq(c, color1);
+                    int dist2 = ColorDistanceSq(c, color2);
+
+                    besti = dist0 < dist1
+                        ? ((dist2 < dist0) ? 2u : 0u)
+                        : ((dist2 < dist1) ? 2u : 1u);
                 }
                 result |= besti << (i * 2);
             }
 
             return result;
         }
 
-        private static void GetMinMaxColor(Color[] block, out Color minCol, out Color maxCol)
+        private static int ColorDistanceSq(Color c0, Color c1)
         {
-            const int INSET_SHIFT = 4;
-            maxCol = Color.TransparentBlack;
-            minCol = Color.White;
+            // Vector distance
+            int r = (c0.R - c1.R) * 5; // Weigh in BT.601 luma coefficients
+            int g = (c0.G - c1.G) * 9;
+            int b = (c0.B - c1.B) * 2;
+            return (r * r) + (g * g) + (b * b);
+        }
 
+        private static void GetMinMaxAlpha(Color[] block, out byte minAlpha, out byte maxAlpha)
+        {
+            byte minA = 255;
+            byte maxA = 0;
             for (int i = 0; i < 16; i++)
             {
-                var col = block[i];
-
-                if (col.A < minCol.A) minCol.A = col.A;
-                if (col.A > maxCol.A) maxCol.A = col.A;
-                if (col.A == 0) continue;
+                byte a = block[i].A;
+                if (a < minA) minA = a;
+                if (a > maxA) maxA = a;
+            }
+            minAlpha = minA;
+            maxAlpha = maxA;
+        }
 
-                if (col.R < minCol.R) minCol.R = col.R;
-                if (col.G < minCol.G) minCol.G = col.G;
-                if (col.B < minCol.B) minCol.B = col.B;
+        private static void GetExtremeColors(Color[] block, out Color color0, out ushort colorBin0, out Color color1, out ushort colorBin1, bool dxt1a)
+        {
+            // Calculate average of colors, skip all colors with Alpha equal to 0
+            bool hasAlpha0 = false;
+            int r = 0, g = 0, b = 0, t = 0;
+            for (int i = 0; i < 16; i++)
+            {
+                if (block[i].A > 0)
+                {
+                    r += block[i].R;
+                    g += block[i].G;
+                    b += block[i].B;
+                    ++t;
+                }
+                else
+                {
+                    hasAlpha0 = true;
+                }
+            }
+            Color avg = (t == 0)
+                ? new Color(0, 0, 0)
+                : new Color(r / t, g / t, b / t);
 
-                if (col.R > maxCol.R) maxCol.R = col.R;
-                if (col.G > maxCol.G) maxCol.G = col.G;
-                if (col.B > maxCol.B) maxCol.B = col.B;
+            // Find color furthest from average
+            int leftDist = 0;
+            int leftIdx = 0;
+            for (int i = 0; i < 16; i++)
```

Comment:
This loop can be combined with the next one. That would skip duplicating the loop overhead, and let you do the fetch from block[i] into local variable once instead of (at least) twice.